hebsafeharbor/identifier/context_enhancement/hebrew_context_aware_enhancer.py

```diff
@@ -0,0 +1,363 @@
+        if not found:
+            raise ValueError(
+                "Did not find word '" + word + "' "
+                                               "in the list of tokens although it "
+                                               "is expected to be found"
+            )
+        return i
+
+    @staticmethod
+    def __add_n_words(
+            index: int,
+            n_words: int,
+            lemmas: List[str],
+            lemmatized_filtered_keywords: List[str],
+            is_backward: bool,
+    ) -> List[str]:
+        """
+        Prepare a string of context words.
+
+        Return a list of words which surrounds a lemma at a given index.
+        The words will be collected only if exist in the filtered array
+
+        :param index: index of the lemma that its surrounding words we want
+        :param n_words: number of words to take
+        :param lemmas: array of lemmas
+        :param lemmatized_filtered_keywords: the array of filtered
+               lemmas from the original sentence,
+        :param is_backward: if true take the preceeding words, if false,
+                            take the successing words
+        """
+        i = index
+        context_words = []
+        # The entity itself is no interest to us...however we want to
+        # consider it anyway for cases were it is attached with no spaces
+        # to an interesting context word, so we allow it and add 1 to
+        # the number of collected words
+
+        # collect at most n words (in lower case)
+        remaining = n_words + 1
+        while 0 <= i < len(lemmas) and remaining > 0:
+            lower_lemma = lemmas[i].lower()
+            if lower_lemma in lemmatized_filtered_keywords:
+                context_words.append(lower_lemma)
+                remaining -= 1
+            i = i - 1 if is_backward else i + 1
+        return context_words
+
+    def __add_n_words_forward(
+            self,
+            index: int,
+            n_words: int,
+            lemmas: List[str],
+            lemmatized_filtered_keywords: List[str],
+    ) -> List[str]:
+        return self.__add_n_words(
+            index, n_words, lemmas, lemmatized_filtered_keywords, False
+        )
+
+    def __add_n_words_backward(
+            self,
+            index: int,
+            n_words: int,
+            lemmas: List[str],
+            lemmatized_filtered_keywords: List[str],
+    ) -> List[str]:
+        return self.__add_n_words(
+            index, n_words, lemmas, lemmatized_filtered_keywords, True
+        )
+
+    @staticmethod
+    def __extract_sentence_containing_position(nlp_artifacts, start):
+        for sent in nlp_artifacts.tokens.sents:
+            if (start >= sent.start_char) & (start <= sent.end_char):
+                return sent.text
+        return ''
+
+    def __find_supportive_context_in_sentence(self, sentence, word, recognizer):
+        terms_offsets = recognizer.context_recognizer(sentence, word, recognizer.allowed_prepositions)
```

Comment:
Are we assuming that all recognizers in the list have the context_recognizer attribute? If not, worth checking if it exists or throwing an exception that this should be added.

Reply:
Added handling